File: android/app/src/main/java/com/laneshadow/ui/atoms/ThemeBadge.kt
package com.laneshadow.ui.atoms

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme

@Composable
fun ThemeBadge(
    modifier: Modifier = Modifier,
    variant: BadgeVariant = BadgeVariant.Default,
    icon: (@Composable () -> Unit)? = null,
    opacity: Float = 1f,
    accessibilityLabel: String? = null,
    content: @Composable RowScope.() -> Unit,
) {
    val theme = LocalLaneShadowTheme.current
    val resolvedOpacity = opacity.coerceIn(0f, 1f)
    val colors = themeBadgeColors(variant = variant, opacity = resolvedOpacity)
    val shape = RoundedCornerShape(theme.radius.full)

    Row(
        modifier =
            modifier
                .defaultMinSize(minHeight = theme.space.lg + theme.space.xs)
                .clip(shape)
                .background(colors.containerColor, shape)
                .then(
                    if (colors.borderColor != null) {
                        Modifier.border(width = 1.dp, color = colors.borderColor, shape = shape)
                    } else {
                        Modifier
                    },
                )
                .padding(horizontal = 10.dp, vertical = 2.dp)
                .then(
                    if (accessibilityLabel != null) {
                        Modifier.semantics { contentDescription = accessibilityLabel }
                    } else {
                        Modifier
                    },
                ),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(theme.space.xs),
    ) {
        if (icon != null) {
            icon()
        }
        Row(
            verticalAlignment = Alignment.CenterVertically,
            content = content,
        )
    }
}

@Composable
fun Badge(
    modifier: Modifier = Modifier,
    variant: BadgeVariant = BadgeVariant.Default,
    icon: (@Composable () -> Unit)? = null,
    opacity: Float = 1f,
    accessibilityLabel: String? = null,
    content: @Composable RowScope.() -> Unit,
) = ThemeBadge(
    modifier = modifier,
    variant = variant,
    icon = icon,
    opacity = opacity,
    accessibilityLabel = accessibilityLabel,
    content = content,
)

enum class BadgeVariant {
    Default,
    Secondary,
    Destructive,
    Outline,
    Success,
    Warning,
    Info,
}

data class ThemeBadgeResolvedColors(
    val containerColor: Color,
    val contentColor: Color,
    val borderColor: Color? = null,
)

@Composable
internal fun themeBadgeColors(
    variant: BadgeVariant,
    opacity: Float = 1f,
): ThemeBadgeResolvedColors {
    val theme = LocalLaneShadowTheme.current
    val resolvedOpacity = opacity.coerceIn(0f, 1f)

    return when (variant) {
        BadgeVariant.Secondary ->
            ThemeBadgeResolvedColors(
                containerColor = theme.colors.secondary.default.copy(alpha = resolvedOpacity),
                contentColor = theme.colors.onSecondary.default,
            )
        BadgeVariant.Destructive ->
            ThemeBadgeResolvedColors(
                containerColor = theme.colors.danger.default.copy(alpha = resolvedOpacity),
                contentColor = theme.colors.onPrimary.default,
            )
        BadgeVariant.Outline ->
            ThemeBadgeResolvedColors(
                containerColor = Color.Transparent,
                contentColor = theme.colors.onSurface.default,
                borderColor = theme.colors.border.default,
            )
        BadgeVariant.Success ->
            ThemeBadgeResolvedColors(
                containerColor = theme.colors.success.default.copy(alpha = resolvedOpacity),
                contentColor = theme.colors.onPrimary.default,
            )
        BadgeVariant.Warning ->
            ThemeBadgeResolvedColors(
                containerColor = theme.colors.warning.default.copy(alpha = resolvedOpacity),
                contentColor = theme.colors.onPrimary.default,
            )
        BadgeVariant.Info ->
            ThemeBadgeResolvedColors(
                containerColor = theme.colors.info.default.copy(alpha = resolvedOpacity),
                contentColor = theme.colors.onPrimary.default,
            )
        BadgeVariant.Default ->
            ThemeBadgeResolvedColors(
                containerColor = theme.colors.primary.default.copy(alpha = resolvedOpacity),
                contentColor = theme.colors.onPrimary.default,
            )
    }
}

File: ios/LaneShadow/Views/Atoms/ThemeBadge.swift
import LaneShadowTheme
import SwiftUI

struct ThemeBadge<Label: View>: View {
    @Environment(\.theme) private var theme

    let variant: ThemeBadgeVariant
    let iconName: String?
    let opacity: Double
    let accessibilityLabel: String?
    @ViewBuilder let label: () -> Label

    init(
        variant: ThemeBadgeVariant = .default,
        iconName: String? = nil,
        opacity: Double = 1,
        accessibilityLabel: String? = nil,
        @ViewBuilder label: @escaping () -> Label
    ) {
        self.variant = variant
        self.iconName = iconName
        self.opacity = opacity
        self.accessibilityLabel = accessibilityLabel
        self.label = label
    }

    var body: some View {
        HStack(spacing: theme.space.xs) {
            if let iconName {
                ThemeIcon(name: iconName, size: theme.space.sm + 2, color: resolvedColors.foreground)
            }

            label()
                .foregroundStyle(resolvedColors.foreground)
        }
        .padding(.horizontal, 10)
        .padding(.vertical, 2)
        .background(resolvedColors.background)
        .overlay {
            Capsule(style: .continuous)
                .stroke(resolvedColors.border ?? .clear, lineWidth: resolvedColors.border == nil ? 0 : 1)
        }
        .clipShape(Capsule(style: .continuous))
        .accessibilityElement(children: .combine)
        .accessibilityLabel(accessibilityLabel ?? variant.rawValue.capitalized)
    }

    private var resolvedColors: ThemeBadgeResolvedColors {
        ThemeBadgeResolvedColors.resolve(theme: theme, variant: variant, opacity: opacity)
    }
}

enum ThemeBadgeVariant: String, CaseIterable {
    case `default`
    case secondary
    case destructive
    case outline
    case success
    case warning
    case info
}

private struct ThemeBadgeResolvedColors {
    let background: Color
    let foreground: Color
    let border: Color?

    static func resolve(theme: Theme, variant: ThemeBadgeVariant, opacity: Double) -> Self {
        let resolvedOpacity = max(0, min(1, opacity))

        switch variant {
        case .default:
            return .init(
                background: theme.colors.primary.default.opacity(resolvedOpacity),
                foreground: theme.colors.onPrimary.default,
                border: nil
            )
        case .secondary:
            return .init(
                background: theme.colors.secondary.default.opacity(resolvedOpacity),
                foreground: theme.colors.onSecondary.default,
                border: nil
            )
        case .destructive:
            return .init(
                background: theme.colors.danger.default.opacity(resolvedOpacity),
                foreground: theme.colors.onPrimary.default,
                border: nil
            )
        case .outline:
            return .init(
                background: .clear,
                foreground: theme.colors.onSurface.default,
                border: theme.colors.border.default
            )
        case .success:
            return .init(
                background: theme.colors.success.default.opacity(resolvedOpacity),
                foreground: theme.colors.onPrimary.default,
                border: nil
            )
        case .warning:
            return .init(
                background: theme.colors.warning.default.opacity(resolvedOpacity),
                foreground: theme.colors.onPrimary.default,
                border: nil
            )
        case .info:
            return .init(
                background: theme.colors.info.default.opacity(resolvedOpacity),
                foreground: theme.colors.onPrimary.default,
                border: nil
            )
        }
    }
}
